File: src/setup/src/managed/Microsoft.NET.HostModel/Bundle/FileEntry.cs
﻿// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using System.IO;

namespace Microsoft.NET.HostModel.Bundle
{
    /// <summary>
    /// FileEntry: Records information about embedded files.
    /// 
    /// The bundle manifest records the following meta-data for each 
    /// file embedded in the bundle:
    /// * Type       (1 byte)
    /// * NameLength (7-bit extension encoding, typically 1 byte)
    /// * Name       ("NameLength" Bytes)
    /// * Offset     (Int64)
    /// * Size       (Int64)
    /// </summary>
    public class FileEntry
    {
        public long Offset;
        public long Size;
        public FileType Type;
        public string RelativePath; // Path of an embedded file, relative to the Bundle source-directory.

        public FileEntry(FileType fileType, string relativePath, long offset, long size)
        {
            Type = fileType;
            RelativePath = relativePath.Replace(Path.DirectorySeparatorChar, Manifest.DirectorySeparatorChar);
            Offset = offset;
            Size = size;
        }

        public void Write(BinaryWriter writer)
        {
            writer.Write(Offset);
            writer.Write(Size);
            writer.Write((byte)Type);
            writer.Write(RelativePath);
        }

        public static FileEntry Read(BinaryReader reader)
        {
            long offset = reader.ReadInt64();
            long size = reader.ReadInt64();
            FileType type = (FileType)reader.ReadByte();
            string fileName = reader.ReadString();
            return new FileEntry(type, fileName, offset, size);
        }

        public override string ToString()
        {
            return String.Format($"{RelativePath} [{Type}] @{Offset} Sz={Size}");
        }
    }
}

File: src/setup/src/corehost/cli/apphost/bundle/file_entry.cpp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

#include "bundle_runner.h"
#include "pal.h"
#include "error_codes.h"
#include "trace.h"
#include "utils.h"

using namespace bundle;

bool file_entry_t::is_valid()
{
    return data.offset > 0 && data.size > 0 &&
        (file_type_t)data.type < file_type_t::__last &&
        data.path_length > 0 && data.path_length <= PATH_MAX;
}

file_entry_t* file_entry_t::read(FILE* stream)
{
    file_entry_t* entry = new file_entry_t();

    // First read the fixed-sized portion of file-entry
    bundle_runner_t::read(&entry->data, sizeof(entry->data), stream);
    if (!entry->is_valid())
    {
        trace::error(_X("Failure processing application bundle; possible file corruption."));
        trace::error(_X("Invalid FileEntry detected."));
        throw StatusCode::BundleExtractionFailure;
    }

    // Read the relative-path, given its length 
    pal::string_t& path = entry->relative_path;
    bundle_runner_t::read_string(path, entry->data.path_length, stream);

    // Fixup the relative-path to have current platform's directory separator.
    if (bundle_dir_separator != DIR_SEPARATOR)
    {
        for (size_t pos = path.find(bundle_dir_separator);
            pos != pal::string_t::npos;
            pos = path.find(bundle_dir_separator, pos))
        {
            path[pos] = DIR_SEPARATOR;
        }
    }

    return entry;
}


